dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ChatStreamingResult.cs

```diff
@@ -15,41 +16,100 @@ namespace Microsoft.SemanticKernel.Connectors.AI.OpenAI.AzureSdk;
+        {
+            if (string.IsNullOrEmpty(role))
+            {
+                role = message.Role.ToString();
+            }
+
+            fullMessage.Append(message.Content);
+        }
+
+        if (fullMessage.Length == 0)
         {
             throw new SKException("Unable to get chat message from stream");
         }
 
-        return new SKChatMessage(chatMessage);
+        return new SKChatMessage(role, fullMessage.ToString());
     }
 
     /// <inheritdoc/>
     public async IAsyncEnumerable<ChatMessageBase> GetStreamingChatMessageAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
     {
-        await foreach (var message in this._choice.GetMessageStreaming(cancellationToken))
+        string role = string.Empty;
+        int i = 0;
+        CompletionsFinishReason? finishReason = null;
+        string? messageId = null;
+        DateTimeOffset? created = null;
+
+        while (i < this._chatUpdates.Count)
         {
-            if (message.FunctionCall is not null || message.Content is { Length: > 0 })
+            var message = this._chatUpdates[i];
+
+            // The role will be passed to further messages as soon it is updated
+            if (string.IsNullOrEmpty(role) && message.Role.HasValue)
+            {
+                role = message.Role.Value.ToString();
+            }
+
+            if (message.FinishReason.HasValue)
+            {
+                finishReason = message.FinishReason.Value;
+            }
+
+            if (!string.IsNullOrEmpty(message.Id))
+            {
+                messageId = message.Id;
+            }
+
+            if (message.Created != default)
+            {
+                created = message.Created;
+            }
+
+            if (message.ContentUpdate is { Length: > 0 })
+            {
+                yield return new SKChatMessage(role, message.ContentUpdate, new() { { nameof(message.ContentUpdate), message.ContentUpdate } });
+            }
+
+            // This part may change to expose the function name and arguments as new properties of a message
+            if (message.FunctionName?.Length > 0 || message.FunctionArgumentsUpdate is { Length: > 0 })
+            {
+                var functionCall = new FunctionCall(message.FunctionName, message.FunctionArgumentsUpdate);
+                yield return new SKChatMessage(role, string.Empty, functionCall, new()
+                {
+                    { nameof(message.FunctionName), message.FunctionName! },
+                    { nameof(message.FunctionArgumentsUpdate), message.FunctionArgumentsUpdate },
+                });
+            }
+
+            i++;
+
+            // Wait for next choice update...
+            while (!this._isStreamEnded && i >= this._chatUpdates.Count)
             {
-                yield return new SKChatMessage(message);
+                await Task.Delay(50, cancellationToken).ConfigureAwait(false);
             }
```

Comment:
This isn't thread-safe nor is it efficient.  We need a better way to actually await for the next message to arrive.